src/aspire/classification/averager2d.py

```diff
@@ -725,6 +750,211 @@ def average(
+        classes = np.atleast_2d(classes)
+        reflections = np.atleast_2d(reflections)
+
+        # Result arrays
+        # These arrays will incrementally store our best alignment.
+        n_classes, n_nbor = classes.shape
+        rotations = np.zeros((n_classes, n_nbor), dtype=self.dtype)
+        dot_products = np.ones((n_classes, n_nbor), dtype=self.dtype) * -np.inf
+        shifts = np.zeros((*classes.shape, 2), dtype=self.dtype)
+
+        # Create a search grid and force initial pair to (0,0)
+        # This is done primarily in case of a tie later, we would prefer unshifted.
+        test_shifts = self._shift_search_grid(
+            self.src.L,
+            self.radius,
+            roll_zero=True,
+            sub_pixel=self.sub_pixel,
+        )
+
+        # Work arrays
+        bs = min(self.batch_size, len(test_shifts))
+        _rotations = np.zeros((bs, n_nbor), dtype=self.dtype)
+        _dot_products = np.ones((bs, n_nbor), dtype=self.dtype) * -np.inf
+        template_images = xp.empty((bs, self.src.L, self.src.L), dtype=self.dtype)
+        _images = xp.empty((n_nbor - 1, self.src.L, self.src.L), dtype=self.dtype)
+
+        for k in trange(n_classes, desc="Rotationally aligning classes"):
+            # We want to locally cache the original images,
+            #  because we will mutate them with shifts in the next loop.
+            #  This avoids recomputing them before each shift
+            # The coefficient for the base images are also computed here.
+            if basis_coefficients is None:
+                original_images = Image(self._cls_images(classes[k], src=self.src))
+            else:
+                original_coef = basis_coefficients[classes[k], :]
+                original_images = self.alignment_basis.evaluate(original_coef)
+
+            _images[:] = xp.asarray(original_images[1:].asnumpy().copy())
+
+            # Handle reflections
+            refl = reflections[k][1:]  # skips original_image 0
+            _images[refl] = xp.flip(_images[refl], axis=-2)
+
+            # Mask off
+            _images[:] = _images[:] * self._mask
+
+            # Convert to polar Fourier
+            pf_images = self._pft.half_to_full(self._pft._transform(_images))
+
+            # Batch over shift search space, updating best results
+            pbar = tqdm(
+                total=len(test_shifts),
+                desc="\tmaximizing over shifts",
+                disable=len(test_shifts) == 1,
+                leave=False,
+            )
+            for start in range(0, len(test_shifts), self.batch_size):
+                end = min(start + self.batch_size, len(test_shifts))
+                bs = end - start  # handle a small last batch
+                batch_shifts = test_shifts[start:end]
+
+                # Shift the base, original_image[0], for each shift in this batch
+                # Note this includes shifting for the zero case
+                template_images[:bs] = xp.asarray(
```

Comment:
This can potentially be sped up by polar Fourier-transforming the base images once, then applying phase shifts (elementwise multiplication in Fourier) on those polar Fourier representations.

Reply:
Sorry for the hold up here.  My first pass at this also didn't work well.  Still working on it.

Reply:
I was able to figure this one out.  For some reason I think the Polar freq grids are XY swapped.  I'll make an issue about checking that, but for now I was able to create `PolarFT.shift` and a unit test as things are.  I will replace the shifting in `averager2d.py` next for the speedup.

Reply:
Ok that's not too bad…